onnxruntime/python/tools/symbolic_shape_infer.py

```diff
@@ -498,6 +498,28 @@ def _onnx_infer_single_node(self, node):
                     if (name in self.initializers_ and name not in self.graph_inputs_)
                 ]
 
+            if node.op_type in [
+                "Add",
+                "Sub",
+                "Mul",
+                "Div",
+                "MatMul",
+                "MatMulInteger",
+                "MatMulInteger16",
+                "Where",
+                "Sum",
+            ]:
+                if node.output[0] in self.known_vi_:
+                    vi = self.known_vi_[node.output[0]]
+                    out_rank = len(get_shape_from_type_proto(vi.type))
+                    in_shapes = [self._get_shape(node, i) for i in range(len(node.input))]
+                    for d in range(
+                        out_rank - (2 if node.op_type in ["MatMul", "MatMulInteger", "MatMulInteger16"] else 0)
```

Comment:
Add some comments why MatMul is 2 here

Reply:
Hi, this code snippet is copied [here](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/symbolic_shape_infer.py#L2680-L2697), this is needed, because within the while loop, self.known_vi_ are always initialized from initialzer or inputs, but some logit will change self.known_vi_ dynamically.

Reply:
I see. Could you move the code to a function to avoid duplicating the code?

Reply:
I can, but naming is too complicated, can you please help me do it, thanks.